File: iOS/Mobile/VideoViewController.m
//
//  VideoViewController.m
//  Mobile
//
//  Created by jkh on 6/28/13.
//  Copyright (c) 2013 Ellucian. All rights reserved.
//

#import "VideoViewController.h"
#import "AppearanceChanger.h"
#import "UIViewController+GoogleAnalyticsTrackerSupport.h"
#import "MBProgressHUD.h"
#import <AVFoundation/AVFoundation.h>

@interface VideoViewController ()

@end

@implementation VideoViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    MBProgressHUD *hud = [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    hud.labelText = NSLocalizedString(@"Loading", @"loading message while waiting for data to load");
    [self.mediaPlayOverlay setHidden:YES];

    self.navigationController.navigationBar.translucent = NO;
    
    self.title = self.module.name;
    self.label.text = [self.module propertyForKey:@"description"];
    if([self.label.text length] == 0) {
        [self.textBackgroundView setHidden:YES];
        [self.label setHidden:YES];
    }
    self.playBackgroundView.backgroundColor = [UIColor primaryColor];
    NSString *urlString = [self.module propertyForKey:@"video"];
    NSURL *fileURL = [NSURL URLWithString:urlString];

    self.moviePlayerViewController = [[MPMoviePlayerViewController alloc] initWithContentURL:fileURL];
    
    dispatch_async(dispatch_get_global_queue(0,0), ^{


        AVURLAsset *asset = [[AVURLAsset alloc] initWithURL:fileURL options:nil];
        AVAssetImageGenerator *generate = [[AVAssetImageGenerator alloc] initWithAsset:asset];
        generate.appliesPreferredTrackTransform = YES;
        NSError *error = NULL;
        CMTime time = CMTimeMake(0, 600);
        CGImageRef imageRef = [generate copyCGImageAtTime:time actualTime:NULL error:&error];
        UIImage *image = [[UIImage alloc] initWithCGImage:imageRef];
        dispatch_async(dispatch_get_main_queue(), ^{
            if(image) {

                    [MBProgressHUD hideHUDForView:self.view animated:YES];
                    [self.mediaPlayOverlay setHidden:NO];
                    [self.imageView setImage:image];
                    self.imageView.userInteractionEnabled = YES;
        
                    UITapGestureRecognizer *tapRecognizer = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(playMovie:)];
                    tapRecognizer.numberOfTapsRequired = 1;
                    [self.imageView addGestureRecognizer:tapRecognizer];
                }
        });
    });
}

-(void) viewDidAppear:(BOOL)animated
{
    [super viewDidAppear:animated];
    [self sendView:@"Video" forModuleNamed:self.module.name];
}

-(void) playMovie:(id)sender
{
    [self sendEventToTracker1WithCategory:kAnalyticsCategoryUI_Action withAction:kAnalyticsActionButton_Press withLabel:@"Play button pressed" withValue:nil forModuleNamed:self.module.name];
    [self.moviePlayerViewController.moviePlayer prepareToPlay];
    [self presentMoviePlayerViewControllerAnimated:self.moviePlayerViewController];
}

@end

File: iOS/Mobile/VideoViewController.swift
//
//  VideoViewController.swift
//  Mobile
//
//  Created by Jason Hocker on 8/21/15.
//  Copyright © 2015 Ellucian Company L.P. and its affiliates. All rights reserved.
//

import Foundation
import AVKit
import AVFoundation

class VideoViewController : UIViewController, UIGestureRecognizerDelegate {
    
    var module : Module?
    
    @IBOutlet var imageView: UIImageView!
    @IBOutlet var textBackgroundView: UIView!
    @IBOutlet var label: UILabel!
    @IBOutlet var videoView: UIView!
    @IBOutlet var mediaPlayButton: UIView!
    
    var assetUrl : NSURL?
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        self.sendView("Video", forModuleNamed: module!.name)
    }
    
    override func viewDidLoad() {
        
        let hud = MBProgressHUD.showHUDAddedTo(self.view, animated: true)
        hud.labelText = NSLocalizedString("Loading", comment: "loading message while waiting for data to load")
        
        
        self.title = module!.name
        if let labelText = self.module?.propertyForKey("description") where labelText.characters.count > 0 {
            label.text = labelText
        } else {
            textBackgroundView.hidden = true
            label.hidden = true
        }
        
        if let urlString = module?.propertyForKey("video") {
            assetUrl = NSURL(string: urlString)
            copyImageToBackground()
        }
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue,
        sender: AnyObject?) {
            sendEventToTracker1WithCategory(kAnalyticsCategoryUI_Action, withAction: kAnalyticsActionButton_Press, withLabel: "Play button pressed", withValue: nil, forModuleNamed: module!.name)
            let destination = segue.destinationViewController as!
            AVPlayerViewController
            if let url = self.assetUrl {
                do {
                    let player = AVPlayer(URL: url)
                    let audioSession = AVAudioSession.sharedInstance()
                    try audioSession.setCategory(AVAudioSessionCategoryPlayback)
                    destination.player = player
                    player.play()
                } catch let error {
                    print(error)
                }
            }
    }
    
    private func copyImageToBackground() {
        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)) {
            let asset: AVAsset = AVAsset(URL: self.assetUrl!)
            let imageGenerator = AVAssetImageGenerator(asset: asset);
            let time = CMTimeMake(0, 600)
            do {
                let imageRef = try imageGenerator.copyCGImageAtTime(time, actualTime: nil)
                let image = UIImage(CGImage: imageRef)
                dispatch_async(dispatch_get_main_queue()) {
                    self.mediaPlayButton.hidden = false
                    MBProgressHUD.hideHUDForView(self.view, animated: true)
                    self.imageView.image = image
                    
                    self.addGestureRecognizer()
                }
            } catch let error as NSError {
                dispatch_async(dispatch_get_main_queue()) {
                    
                    MBProgressHUD.hideHUDForView(self.view, animated: true)
                    let alertController = UIAlertController(title: NSLocalizedString("Error Loading Video", comment: "title when error loading video"), message: error.localizedDescription, preferredStyle: .Alert)
                    
                    let OKAction = UIAlertAction(title: "OK", style: .Default, handler: nil)
                    alertController.addAction(OKAction)
                    
                    self.presentViewController(alertController, animated: true, completion: nil)
                }
            } catch {
                
            }
        }
    }
    
    func addGestureRecognizer() {
        let recognizer = UITapGestureRecognizer(target: self, action:Selector("play:"))
        recognizer.delegate = self
        self.videoView.addGestureRecognizer(recognizer)
    }
    
    func play(recognizer: UITapGestureRecognizer) {
        self.performSegueWithIdentifier("play", sender: nil)
    }
}
